File: core/src/main/java/heroes/journey/registries/Registrable.java
package heroes.journey.registries;

import lombok.Getter;

@Getter
public class Registrable {

    private final String id;
    private final String name;

    public Registrable(String id, String name) {
        this.id = id;
        this.name = name;
    }

    public Registrable(String id) {
        this.id = id;
        this.name = id;
    }

    public String toString() {
        return name == null ? id : name;
    }
}

File: core/src/main/java/heroes/journey/registries/Registrable.kt
package heroes.journey.registries

import heroes.journey.utils.Lang

abstract class Registrable(val id: String) {
    init {
        require(id.isNotBlank()) { "Renderable id must not be blank or empty" }
    }

    open fun getName(): String {
        return Lang.get("${id}_name")
    }

    open fun getDescription(): String {
        return Lang.get("${id}_description")
    }

    override fun toString(): String {
        return id
    }

    abstract fun register(): Registrable
}
